File: casino/core/wellness.py
"""Optional session wellness reminders, break controls, and neutral session summaries. (#167)

Everything here is opt-in and off by default. The module deliberately offers no reward, streak,
bonus, or token movement for engaging with a reminder, and it never evaluates a player's results:
summaries report committed play-token totals and elapsed time as plain numbers, and every piece of
player-visible wording lives in the shell resources so the neutral-copy rule is enforced against
shipped text rather than against strings scattered through Python.

Reminder intervals are bounded at both ends. A floor exists so the product cannot be configured
into countdown pressure, and a ceiling exists so a reminder a player enabled still actually arrives.
"""

# Import the shared UTC formatter for contract-compatible timestamps.
from casino.core.clock import utc_now
# Import the authoritative ledger so summaries report committed movements only.
from casino.core import ledger
# Import the shared self-history ceiling so every personal ledger read stays identically bounded.
from casino.core import self_history as activity
# Import the configured storage provider for atomic document persistence.
from casino.core.storage import get_storage_provider
# Import standard bounded application errors.
from casino.errors import ConflictError, ValidationError

# Name the wellness document owned by this module.
WELLNESS_DOCUMENT_KEY = "settings/wellness"
# Enumerate the only wellness fields a caller may ever submit.
ALLOWED_FIELDS = frozenset({"enabled", "reminder_interval_minutes", "break_reminder_enabled"})
# Keep reminders off until a player explicitly opts in.
DEFAULT_ENABLED = False
# Use a calm default cadence rather than a frequent, pressuring one.
DEFAULT_INTERVAL_MINUTES = 30
# Refuse an interval short enough to function as countdown pressure.
MIN_INTERVAL_MINUTES = 10
# Refuse an interval so long that an enabled reminder would never arrive.
MAX_INTERVAL_MINUTES = 240
# Build the default wellness record for a player who has never configured reminders.
def default_wellness() -> dict:
    # Return the opt-in defaults with the initial revision.
    return {"enabled": DEFAULT_ENABLED, "reminder_interval_minutes": DEFAULT_INTERVAL_MINUTES, "break_reminder_enabled": DEFAULT_ENABLED, "revision": 0, "updated_at": None}


# Build the empty persisted document shape.
def _default_document() -> dict:
    # Hold one entry per user id.
    return {"users": {}}


# Normalize any stored record into the published shape without trusting persisted types.
def _normalize(record) -> dict:
    # Start from the opt-in defaults so partial or malformed state still yields a valid answer.
    wellness = default_wellness()
    # Leave the defaults untouched when the stored value is not a record.
    if not isinstance(record, dict):
        # Return the defaults for malformed persisted state.
        return wellness
    # Inspect both stored switches without allowing truthy malformed values to opt a player in.
    for flag in ("enabled", "break_reminder_enabled"):
        # Adopt only a strict stored boolean.
        if isinstance(record.get(flag), bool):
            # Preserve the valid stored choice.
            wellness[flag] = record[flag]
    # Adopt a stored interval only when it is still inside the accepted bounds.
    stored = record.get("reminder_interval_minutes")
    # Require a usable in-range integer before trusting a persisted cadence.
    if isinstance(stored, int) and MIN_INTERVAL_MINUTES <= stored <= MAX_INTERVAL_MINUTES:
        # Preserve the configured cadence.
        wellness["reminder_interval_minutes"] = stored
    # Adopt a stored revision only when it is a usable non-negative integer.
    if isinstance(record.get("revision"), int) and record["revision"] >= 0:
        # Preserve the concurrency revision.
        wellness["revision"] = record["revision"]
    # Adopt a stored timestamp only when it is a string.
    if isinstance(record.get("updated_at"), str):
        # Preserve the last update time.
        wellness["updated_at"] = record["updated_at"]
    # Return the normalized published record.
    return wellness


# Resolve the durable subject for one authenticated session without trusting caller input.
def _subject(user) -> str:
    # Read the session-bound durable identity.
    user_id = str((user or {}).get("user_id") or "")
    # Refuse to operate without an authenticated subject.
    if not user_id:
        # Fail closed rather than defaulting to a shared record.
        raise ValidationError("Session wellness requires an authenticated session", {"reason": "no_subject"})
    # Return the session-derived subject.
    return user_id


# Report whether one subject is a disposable guest trial rather than a persistent account.
def _is_guest(user) -> bool:
    # Treat the accepted guest markers as the single disposable-session signal.
    return str((user or {}).get("role") or "") == "guest" or bool((user or {}).get("guest_analytics_id"))


# Read the wellness preferences bound to the authenticated session.
def read_wellness(user) -> dict:
    # Derive the subject from the session only.
    subject = _subject(user)
    # Return non-persisted defaults for a disposable guest trial.
    if _is_guest(user):
        # Publish explicit persistence metadata so a client never promises durable guest settings.
        return {**default_wellness(), "min_interval_minutes": MIN_INTERVAL_MINUTES, "max_interval_minutes": MAX_INTERVAL_MINUTES, "persisted": False}
    # Read the wellness document through the configured provider.
    document = get_storage_provider().read_document(WELLNESS_DOCUMENT_KEY, _default_document)
    # Tolerate a malformed document by treating it as empty rather than failing the read.
    users = document.get("users") if isinstance(document, dict) else None
    # Resolve this subject's stored record when the container is usable.
    record = users.get(subject) if isinstance(users, dict) else None
    # Publish the normalized record together with the accepted cadence bounds.
    return {**_normalize(record), "min_interval_minutes": MIN_INTERVAL_MINUTES, "max_interval_minutes": MAX_INTERVAL_MINUTES, "persisted": True}


# Validate one caller-supplied wellness patch against the field allowlist and cadence bounds.
def _validated_patch(patch) -> dict:
    # Require an object body.
    if not isinstance(patch, dict):
        # Reject a non-object payload.
        raise ValidationError("Wellness update must be an object", {"reason": "malformed"})
    # Reject any field outside the allowlist so unknown keys can never reach storage.
    unknown = sorted(set(patch) - ALLOWED_FIELDS - {"revision"})
    # Fail closed when the caller submitted an unsupported field.
    if unknown:
        # Name only the rejected field keys, never their values.
        raise ValidationError("Wellness update contains unsupported fields", {"reason": "unsupported_fields", "fields": unknown})
    # Collect only the validated changes.
    changes = {}
    # Validate both switches when supplied.
    for flag in ("enabled", "break_reminder_enabled"):
        # Skip a switch the caller did not send.
        if flag in patch:
            # Require a strict boolean so a truthy string cannot enable a reminder.
            if not isinstance(patch[flag], bool):
                # Reject a non-boolean switch.
                raise ValidationError("Wellness switches must be booleans", {"reason": "malformed_switch", "field": flag})
            # Accept the boolean switch.
            changes[flag] = patch[flag]
    # Validate the reminder cadence when supplied.
    if "reminder_interval_minutes" in patch:
        # Read the requested cadence.
        interval = patch["reminder_interval_minutes"]
        # Refuse a non-integer, sub-floor, or above-ceiling cadence so reminders cannot become pressure.
        if not isinstance(interval, int) or isinstance(interval, bool) or not (MIN_INTERVAL_MINUTES <= interval <= MAX_INTERVAL_MINUTES):
            # Publish the accepted bounds so a client can correct itself.
            raise ValidationError("Reminder interval is outside the accepted range", {"reason": "interval_out_of_range", "min": MIN_INTERVAL_MINUTES, "max": MAX_INTERVAL_MINUTES})
        # Accept the bounded cadence.
        changes["reminder_interval_minutes"] = interval
    # Require at least one real change so an empty write cannot advance the revision.
    if not changes:
        # Reject a patch that carries no supported field.
        raise ValidationError("Wellness update contains no supported changes", {"reason": "empty_update"})
    # Return the validated changes.
    return changes


# Apply a validated wellness change for the authenticated session.
def update_wellness(user, patch) -> dict:
    # Derive the subject from the session only.
    subject = _subject(user)
    # Validate the caller payload before touching storage.
    changes = _validated_patch(patch)
    # Read the caller's expected revision when it supplied one for optimistic concurrency.
    expected = patch.get("revision") if isinstance(patch, dict) else None
    # Keep disposable guest changes non-durable and explicit about that lifecycle.
    if _is_guest(user):
        # Return only the validated session-local values without creating a durable account record.
        return {**default_wellness(), **changes, "min_interval_minutes": MIN_INTERVAL_MINUTES, "max_interval_minutes": MAX_INTERVAL_MINUTES, "persisted": False}
    # Hold the resulting record so it can be published after the atomic mutation.
    resulting = {}

    # Mutate only this subject's entry inside the shared document.
    def mutate(document):
        # Replace a malformed document rather than failing every future write.
        if not isinstance(document, dict):
            # Start from the empty shape.
            document = _default_document()
        # Replace a malformed container while preserving the surrounding document.
        if not isinstance(document.get("users"), dict):
            # Reset only the container this module owns.
            document["users"] = {}
        # Normalize the current stored record for this subject.
        current = _normalize(document["users"].get(subject))
        # Reject a stale revision so two sessions cannot silently overwrite one another.
        if expected is not None and expected != current["revision"]:
            # Use the standard conflict envelope without revealing stored values.
            raise ConflictError("Wellness settings were updated by another session")
        # Build the next record from the current values and the validated changes.
        nxt = {**current, **changes, "revision": current["revision"] + 1, "updated_at": utc_now()}
        # Store the next record for this subject only.
        document["users"][subject] = nxt
        # Publish the stored record to the caller.
        resulting.update(nxt)
        # Return the mutated document for atomic persistence.
        return document

    # Persist the change atomically through the configured provider.
    get_storage_provider().update_document(WELLNESS_DOCUMENT_KEY, mutate, _default_document)
    # Publish the stored record together with the accepted cadence bounds.
    return {**resulting, "min_interval_minutes": MIN_INTERVAL_MINUTES, "max_interval_minutes": MAX_INTERVAL_MINUTES, "persisted": True}


# Summarize the authenticated session's own committed play-token activity without evaluating it.
def session_summary(user, *, since: str = "") -> dict:
    # Derive the subject from the session only, never from a caller-supplied identifier.
    _subject(user)
    # Resolve the ledger subject bound to this session.
    player_id = str((user or {}).get("player_id") or "")
    # Publish an explicit empty summary when the session has no ledger subject.
    if not player_id:
        # Return a stable zeroed envelope rather than another player's totals.
        return {"movements": 0, "staked": 0.0, "returned": 0.0, "net": 0.0, "since": str(since or ""), "play_tokens_only": True}
    # Read only this player's committed movements.
    rows = ledger.read_recent(player_id, activity.LEDGER_READ_CEILING)
    # Drop any row not bound to this subject so a provider change can never leak another player.
    owned = [row for row in rows if str(row.get("player_id") or "") == player_id]
    # Apply the optional start boundary using the committed timestamps.
    bounded = [row for row in owned if not since or str(row.get("ts") or "") >= str(since)] if since else owned
    # Total the committed outgoing movements as staked play tokens.
    staked = sum(-float(row.get("amount") or 0) for row in bounded if float(row.get("amount") or 0) < 0)
    # Total the committed incoming movements as returned play tokens.
    returned = sum(float(row.get("amount") or 0) for row in bounded if float(row.get("amount") or 0) > 0)
    # Publish plain committed totals with no evaluative language or judgement.
    return {"movements": len(bounded), "staked": round(staked, 2), "returned": round(returned, 2), "net": round(returned - staked, 2), "since": str(since or ""), "play_tokens_only": True}


# Record that a player acknowledged a reminder without granting anything in return.
def acknowledge_reminder(user) -> dict:
    # Derive the subject so an unauthenticated caller cannot record an acknowledgement.
    _subject(user)
    # Return only the acknowledgement time; no token, streak, bonus, or reward is ever issued.
    return {"acknowledged_at": utc_now(), "reward_granted": False}

File: web/core/wellness.js
// Copyright 2026 Andrei Vorobiev and Virtual Casino Simulator contributors
// SPDX-License-Identifier: Apache-2.0
// Keep the smallest allowed reminder interval aligned with the server contract. (WELL-001)
export const MIN_WELLNESS_INTERVAL = 10;
// Keep the largest allowed reminder interval aligned with the server contract. (WELL-001)
export const MAX_WELLNESS_INTERVAL = 240;

// Normalize one server timestamp without allowing an invalid value to create an immediate reminder loop.
function sessionStartMilliseconds(session, now) {
  // Prefer the public session descriptor used across reloads and route changes.
  const raw = session?.session?.issued_at || session?.session?.created_at || session?.session_status?.issued_at;
  // Parse the server-owned timestamp into a wall-clock boundary.
  const parsed = Date.parse(String(raw || ''));
  // Fall back to the current clock only when an older compatible payload has no usable timestamp.
  return Number.isFinite(parsed) ? parsed : now();
}

// Read one session-local presentation record without treating corrupt storage as an opt-in choice.
function readLocalState(storage, key) {
  // Parse the bounded record inside a failure boundary for disabled or unavailable session storage.
  try {
    // Decode only the current authenticated session's opaque timing record.
    const record = JSON.parse(storage?.getItem(key) || '{}');
    // Return strict primitive values so malformed records cannot pause or skip reminders.
    return { lastSlot: Number.isInteger(record.lastSlot) && record.lastSlot >= 0 ? record.lastSlot : 0, paused: record.paused === true, intervalMinutes: Number.isInteger(record.intervalMinutes) && record.intervalMinutes >= MIN_WELLNESS_INTERVAL && record.intervalMinutes <= MAX_WELLNESS_INTERVAL ? record.intervalMinutes : null };
  // Degrade to an unpaused, undisplayed session when browser storage is unavailable.
  } catch (_) {
    // Preserve opt-in server authority while discarding only corrupt local presentation state.
    return { lastSlot: 0, paused: false, intervalMinutes: null };
  }
}

// Persist only timing presentation state, never player identity, amounts, or API results.
function writeLocalState(storage, key, state) {
  // Keep storage failure from interrupting gameplay or disabling the visible control.
  try {
    // Serialize the two bounded values needed to prevent duplicate reload reminders.
    storage?.setItem(key, JSON.stringify({ lastSlot: state.lastSlot, paused: state.paused, intervalMinutes: state.intervalMinutes }));
  // Ignore disabled browser storage because the active in-memory schedule remains safe.
  } catch (_) { /* Session storage is an optional reload-deduplication aid. */ }
}

// Create the persistent every-game session wellness controller. (WELL-001, WELL-002, issue #167)
export function createWellnessController(options = {}) {
  // Bind the authenticated API helper without granting the controller any other route authority.
  const apiClient = options.apiClient;
  // Bind the persistent document that owns the topbar control and native dialog.
  const documentRef = options.documentRef || globalThis.document;
  // Bind the current browser window for visibility, timers, and session-local storage.
  const windowRef = options.windowRef || globalThis.window;
  // Bind localization through the shell so the controller owns no resource-loading lifecycle.
  const translate = options.translate || ((key, values) => key.replace(/\{(\w+)\}/g, (_, name) => String(values?.[name] ?? '')));
  // Bind play-token formatting through the existing shell formatter.
  const formatTokens = options.formatTokens || (value => String(value));
  // Bind the wall clock so deterministic tests can cross reminder boundaries without sleeping.
  const now = options.now || (() => Date.now());
  // Bind the timer scheduler so every pending wakeup can be cancelled at session teardown.
  const scheduleTimer = options.setTimer || ((callback, delay) => windowRef.setTimeout(callback, delay));
  // Bind timer cancellation to the same injected clock domain.
  const cancelTimer = options.clearTimer || (handle => windowRef.clearTimeout(handle));
  // Cache the persistent topbar button supplied by index.html.
  const openButton = documentRef?.getElementById('wellness-open');
  // Cache the native dialog supplied by index.html.
  const dialog = documentRef?.getElementById('wellness-dialog');
  // Cache the settings form so server writes remain explicit submit actions.
  const form = documentRef?.getElementById('wellness-form');
  // Track the current controller generation across login, logout, and session replacement.
  let generation = 0;
  // Hold the active server wellness record under opt-in defaults.
  let settings = { enabled: false, break_reminder_enabled: false, reminder_interval_minutes: 30, revision: 0, persisted: false };
  // Hold the last authoritative neutral summary for locale-only rerenders.
  let summary = null;
  // Hold the server-owned session start used for absolute, drift-free cadence boundaries.
  let startedAt = 0;
  // Hold the opaque per-session storage key used only for presentation deduplication.
  let localKey = '';
  // Hold reload-stable pause and last-presented-slot state.
  let localState = { lastSlot: 0, paused: false, intervalMinutes: null };
  // Hold the only pending reminder wakeup.
  let timer = null;
  // Remember a reminder deferred while the document was hidden.
  let pendingVisibleReminder = false;
  // Preserve the invoking control so native dialog closure restores useful keyboard focus.
  let restoreFocus = null;

  // Return one persistent dialog node by id without caching stale route-owned elements.
  const node = id => documentRef?.getElementById(id);

  // Translate a shell wellness string with the established formatting contract.
  const copy = (key, values = {}) => translate(`wellness.${key}`, values);

  // Cancel the one scheduled cadence wakeup before replacing session or settings state.
  function clearSchedule() {
    // Cancel only when a real timer handle is retained.
    if (timer !== null) cancelTimer(timer);
    // Clear the handle so repeated teardown is idempotent.
    timer = null;
  }

  // Compute whole elapsed minutes from the server-owned session boundary.
  function elapsedMinutes() {
    // Clamp clock rollback to zero so no negative or pressure-style time appears.
    return Math.max(0, Math.floor((now() - startedAt) / 60000));
  }

  // Render settings, summary, and action state without changing cadence ownership.
  function render() {
    // Update the persistent topbar label for the active locale.
    if (openButton) { openButton.textContent = copy('open'); openButton.setAttribute('aria-label', copy('open')); }
    // Stop after persistent-control localization when the optional dialog markup is absent.
    if (!dialog) return;
    // Localize the dialog eyebrow alongside every other persistent-shell label.
    node('wellness-eyebrow').textContent = copy('eyebrow');
    // Localize the dialog heading and explanatory copy.
    node('wellness-title').textContent = copy('title');
    // Explain the off-by-default and player-controlled behavior.
    node('wellness-copy').textContent = copy('copy');
    // Publish elapsed time as neutral information rather than a countdown.
    node('wellness-elapsed').textContent = copy('elapsed', { minutes: elapsedMinutes() });
    // Localize every settings label and bounded-range hint.
    node('wellness-enable-label').textContent = copy('enable');
    // Localize the optional stopping-point suggestion control.
    node('wellness-break-label').textContent = copy('breakEnable');
    // Localize the cadence field label.
    node('wellness-interval-label').textContent = copy('interval');
    // Publish the server-aligned allowed cadence range.
    node('wellness-interval-help').textContent = copy('intervalRange', { min: MIN_WELLNESS_INTERVAL, max: MAX_WELLNESS_INTERVAL });
    // Localize explicit durable save behavior.
    node('wellness-save').textContent = copy('save');
    // Localize the neutral summary heading.
    node('wellness-summary-title').textContent = copy('summaryTitle');
    // Localize session-local pause or resume according to current state.
    node('wellness-pause').textContent = copy(localState.paused ? 'resume' : 'pause');
    // Localize the durable stop action without urgency.
    node('wellness-stop').textContent = copy('stop');
    // Localize the ordinary close action.
    node('wellness-dismiss').textContent = copy('dismiss');
    // Keep the screen-reader name synchronized with the visible localized close action.
    node('wellness-dismiss').setAttribute('aria-label', copy('dismiss'));
    // Synchronize the elapsed-reminder checkbox with the authoritative server record.
    node('wellness-enabled').checked = settings.enabled === true;
    // Synchronize the stopping-point suggestion checkbox with the authoritative server record.
    node('wellness-break-enabled').checked = settings.break_reminder_enabled === true;
    // Synchronize the bounded cadence field with the authoritative server record.
    node('wellness-interval').value = String(settings.reminder_interval_minutes || 30);
    // Disable pause and stop when elapsed reminders are already durably disabled.
    node('wellness-pause').disabled = settings.enabled !== true;
    // Disable stop when there is no enabled reminder to stop.
    node('wellness-stop').disabled = settings.enabled !== true;
    // Build neutral summary rows only from the latest server-owned result.
    const rows = summary ? [copy('summaryMovements', { count: summary.movements }), copy('summaryStaked', { amount: formatTokens(summary.staked) }), copy('summaryReturned', { amount: formatTokens(summary.returned) }), copy('summaryNet', { amount: formatTokens(summary.net) })] : [copy('summaryLoading')];
    // Replace the summary list with text-only rows so API data cannot become markup.
    node('wellness-summary').replaceChildren(...rows.map(value => { const item = documentRef.createElement('li'); item.textContent = value; return item; }));
    // Keep the play-token disclaimer visible beside every summary.
    node('wellness-summary-note').textContent = copy('summaryNote');
  }

  // Schedule the next absolute interval boundary without accumulating setTimeout drift.
  function scheduleNext() {
    // Cancel the prior wakeup before deriving a replacement.
    clearSchedule();
    // Stop when the user has not opted in or has paused reminders for this login session.
    if (settings.enabled !== true || localState.paused || !startedAt) return;
    // Convert the validated server cadence into milliseconds.
    const intervalMs = Number(settings.reminder_interval_minutes) * 60000;
    // Determine the currently elapsed complete cadence slot.
    const elapsedSlot = Math.max(0, Math.floor((now() - startedAt) / intervalMs));
    // Open a due reminder once when the current slot was not already presented.
    if (elapsedSlot >= 1 && elapsedSlot > localState.lastSlot) {
      // Defer presentation while hidden so background tabs never steal focus.
      if (documentRef.visibilityState === 'hidden') { pendingVisibleReminder = true; return; }
      // Present the due reminder through the same manual dialog surface.
      void open('reminder', elapsedSlot);
      // Stop because open schedules the following boundary after recording this slot.
      return;
    }
    // Schedule the first boundary strictly after both elapsed and last-presented slots.
    const nextSlot = Math.max(1, elapsedSlot + 1, localState.lastSlot + 1);
    // Bound browser delay while retaining an immediate recheck for very long sessions.
    const delay = Math.max(0, Math.min(startedAt + nextSlot * intervalMs - now(), 2147483647));
    // Wake once to re-evaluate visibility, settings, pause, and absolute elapsed time.
    timer = scheduleTimer(() => { timer = null; scheduleNext(); }, delay);
  }

  // Refresh the neutral active-session summary without inventing fallback totals.
  async function refreshSummary(capturedGeneration) {
    // Read only the authenticated session-owned additive summary route.
    const result = await apiClient('/api/v2/me/wellness/summary');
    // Ignore a response that belongs to a replaced or logged-out session.
    if (capturedGeneration !== generation) return false;
    // Adopt the complete server projection for text-only rendering.
    summary = result;
    // Repaint current locale and values without affecting timer state.
    render();
    // Signal that the exact current session adopted the response.
    return true;
  }

  // Open the shared dialog manually or for one due cadence slot.
  async function open(reason = 'manual', dueSlot = 0) {
    // Stop when no authenticated controller generation owns the persistent control.
    if (!generation || !dialog) return;
    // Preserve the current focus only when it is a real document element.
    restoreFocus = documentRef.activeElement instanceof windowRef.HTMLElement ? documentRef.activeElement : openButton;
    // Record a due slot before asynchronous summary I/O so reload cannot duplicate it.
    if (reason === 'reminder' && dueSlot > localState.lastSlot) { localState.lastSlot = dueSlot; pendingVisibleReminder = false; writeLocalState(windowRef.sessionStorage, localKey, localState); }
    // Capture the current generation across the summary request.
    const capturedGeneration = generation;
    // Clear stale status before announcing this explicit dialog state.
    node('wellness-message').textContent = reason === 'reminder' ? copy('reminder', { minutes: elapsedMinutes() }) : '';
    // Add the optional neutral stopping-point suggestion only when enabled.
    node('wellness-break-suggestion').textContent = reason === 'reminder' && settings.break_reminder_enabled ? copy('breakSuggestion') : '';
    // Render immediately so manual controls never wait on the summary endpoint.
    render();
    // Open the native modal once and let the browser own focus trapping and Escape.
    if (!dialog.open) dialog.showModal();
    // Move focus to the heading for a stable screen-reader entry point.
    node('wellness-title').focus();
    // Refresh totals while preserving the already-open usable settings UI.
    try { await refreshSummary(capturedGeneration); } catch (_) { if (capturedGeneration === generation) { summary = null; node('wellness-message').textContent = copy('summaryError'); render(); } }
    // Schedule the next absolute boundary after current presentation settles.
    if (capturedGeneration === generation) scheduleNext();
  }

  // Close the dialog and restore focus to its exact still-connected invoking control.
  function close() {
    // Close only an actually open native dialog.
    if (dialog?.open) dialog.close();
    // Restore focus only when the original element still belongs to this document.
    if (restoreFocus?.isConnected && !restoreFocus.disabled) restoreFocus.focus();
    // Drop the element reference so later route replacement cannot retain it.
    restoreFocus = null;
  }

  // Persist explicit settings form changes through optimistic concurrency.
  async function save(event) {
    // Keep the native dialog open while the bounded API request settles.
    event?.preventDefault?.();
    // Capture the active session generation across the write.
    const capturedGeneration = generation;
    // Disable repeated submission until the exact authoritative response returns.
    node('wellness-save').disabled = true;
    // Clear stale result copy before the new request.
    node('wellness-message').textContent = '';
    // Submit only the allowlisted server wellness fields.
    try {
      // Request one explicit opt-in configuration update.
      const result = await apiClient('/api/v2/me/wellness', { method: 'PATCH', body: { enabled: node('wellness-enabled').checked, break_reminder_enabled: node('wellness-break-enabled').checked, reminder_interval_minutes: Number(node('wellness-interval').value), revision: settings.revision } });
      // Ignore a response belonging to a cleared or replaced login session.
      if (capturedGeneration !== generation) return;
      // Remember the prior cadence so a changed interval cannot reuse incompatible slot numbering.
      const priorInterval = Number(settings.reminder_interval_minutes);
      // Adopt only the exact returned server record.
      settings = result.wellness;
      // Mark the elapsed position under a newly chosen cadence so saving cannot trigger or skip historical slots.
      if (Number(settings.reminder_interval_minutes) !== priorInterval) localState.lastSlot = Math.max(0, Math.floor((now() - startedAt) / (Number(settings.reminder_interval_minutes) * 60000)));
      // Bind subsequent reload deduplication to the cadence that defined the recorded slot.
      localState.intervalMinutes = Number(settings.reminder_interval_minutes);
      // Clear session-local pause when a new explicit configuration is saved.
      localState.paused = false;
      // Persist the local presentation state without exposing the settings payload.
      writeLocalState(windowRef.sessionStorage, localKey, localState);
      // Publish a calm saved acknowledgement.
      node('wellness-message').textContent = copy('saved');
      // Repaint and reschedule from the unchanged server session boundary.
      render(); scheduleNext();
    // Keep the prior authoritative state when validation, conflict, or transport fails.
    } catch (_) {
      // Publish one localized recoverable error without server detail.
      node('wellness-message').textContent = copy('saveError');
      // Restore the exact prior settings controls.
      render();
    // Always restore submission availability for the current dialog generation.
    } finally {
      // Re-enable only when this session still owns the controller.
      if (capturedGeneration === generation) node('wellness-save').disabled = false;
    }
  }

  // Pause or resume reminders only for the current login session.
  function togglePause() {
    // Stop when durable settings are already disabled.
    if (settings.enabled !== true) return;
    // Toggle the strictly session-local reminder suspension.
    localState.paused = !localState.paused;
    // Persist the pause across same-session reload without changing server preferences.
    writeLocalState(windowRef.sessionStorage, localKey, localState);
    // Publish the resulting state neutrally.
    node('wellness-message').textContent = copy(localState.paused ? 'paused' : 'resumed');
    // Repaint the action label and replace the schedule.
    render(); scheduleNext();
  }

  // Durably stop reminders through the same preference route as the settings form.
  async function stop() {
    // Stop when no enabled reminder exists.
    if (settings.enabled !== true) return;
    // Capture the active session before the durable write.
    const capturedGeneration = generation;
    // Lock both mutation controls while the exact request settles.
    node('wellness-stop').disabled = true; node('wellness-save').disabled = true;
    // Submit only the disabling switch and optimistic revision.
    try {
      // Request a durable opt-out without altering cadence or suggestion preference.
      const result = await apiClient('/api/v2/me/wellness', { method: 'PATCH', body: { enabled: false, revision: settings.revision } });
      // Ignore late success after logout or session replacement.
      if (capturedGeneration !== generation) return;
      // Adopt the exact returned opt-out record.
      settings = result.wellness;
      // Clear session-local pause now that the server switch is authoritative.
      localState.paused = false;
      // Persist deduplication state while leaving no amounts or identity in storage.
      writeLocalState(windowRef.sessionStorage, localKey, localState);
      // Cancel pending reminders immediately.
      clearSchedule(); pendingVisibleReminder = false;
      // Publish a calm durable-stop acknowledgement.
      node('wellness-message').textContent = copy('stopped');
      // Repaint disabled action state.
      render();
    // Keep reminders active when the server did not confirm the opt-out.
    } catch (_) {
      // Publish one recoverable error without claiming reminders stopped.
      node('wellness-message').textContent = copy('saveError');
      // Resume the prior authoritative schedule.
      render(); scheduleNext();
    // Restore settings submission for the current session.
    } finally {
      // Re-enable save when the same session still owns this controller.
      if (capturedGeneration === generation) node('wellness-save').disabled = false;
    }
  }

  // Start the controller from one authenticated server session.
  async function start(session) {
    // Advance generation first so every prior response loses ownership.
    generation += 1;
    // Capture the new generation across the initial settings request.
    const capturedGeneration = generation;
    // Cancel old wakeups and close any old-session dialog before reusing persistent nodes.
    clearSchedule(); close(); pendingVisibleReminder = false; summary = null;
    // Derive the absolute session boundary from public server metadata.
    startedAt = sessionStartMilliseconds(session, now);
    // Build one non-identifying session-local storage key from the server timestamp.
    localKey = `casino.wellness.v1.${startedAt}`;
    // Restore only pause and last-presented cadence slot for this exact login session.
    localState = readLocalState(windowRef.sessionStorage, localKey);
    // Keep the control hidden until the authenticated settings read succeeds.
    if (openButton) openButton.hidden = true;
    // Read the exact authenticated user's opt-in settings.
    try {
      // Call only the additive wellness preference route.
      const result = await apiClient('/api/v2/me/wellness');
      // Ignore a response after logout or another login replacement.
      if (capturedGeneration !== generation) return false;
      // Adopt the returned server record under strict booleans and validated cadence.
      settings = result.wellness;
      // Rebase a slot retained under another tab's old cadence without presenting historical intervals.
      if (localState.intervalMinutes !== null && localState.intervalMinutes !== Number(settings.reminder_interval_minutes)) localState.lastSlot = Math.max(0, Math.floor((now() - startedAt) / (Number(settings.reminder_interval_minutes) * 60000)));
      // Record the authoritative cadence used to interpret the current slot number.
      localState.intervalMinutes = Number(settings.reminder_interval_minutes);
      // Persist cadence binding while retaining only bounded presentation metadata.
      writeLocalState(windowRef.sessionStorage, localKey, localState);
      // Reveal the every-game control for this authenticated shell.
      if (openButton) openButton.hidden = false;
      // Render the active locale and schedule the next absolute boundary.
      render(); scheduleNext();
      // Signal successful ownership-aware start.
      return true;
    // Keep the optional feature unavailable when an older or failed server lacks the route.
    } catch (_) {
      // Hide the control and leave gameplay unaffected.
      if (capturedGeneration === generation && openButton) openButton.hidden = true;
      // Report optional-controller unavailability without throwing through authenticated entry.
      return false;
    }
  }

  // Dispose all session-owned work at logout, expiry, or replacement.
  function dispose() {
    // Advance generation so every outstanding request becomes stale.
    generation += 1;
    // Cancel pending cadence work and hidden-tab delivery.
    clearSchedule(); pendingVisibleReminder = false;
    // Close the native dialog before authenticated shell identity is cleared.
    close();
    // Hide the authenticated-only persistent control.
    if (openButton) openButton.hidden = true;
    // Clear in-memory API projections without altering durable user settings.
    settings = { enabled: false, break_reminder_enabled: false, reminder_interval_minutes: 30, revision: 0, persisted: false }; summary = null; startedAt = 0; localKey = '';
  }

  // Deliver one deferred reminder only after the document becomes visible again.
  function handleVisibility() {
    // Stop unless one due reminder is waiting and the current session still owns the controller.
    if (!pendingVisibleReminder || documentRef.visibilityState === 'hidden' || !generation) return;
    // Re-evaluate the exact current cadence slot before presentation.
    const intervalMs = Number(settings.reminder_interval_minutes) * 60000;
    // Derive the due slot from the absolute server-owned boundary.
    const slot = Math.max(1, Math.floor((now() - startedAt) / intervalMs));
    // Present once through the normal reminder path.
    void open('reminder', slot);
  }

  // Bind the persistent manual-open control once for the application lifetime.
  if (openButton) openButton.addEventListener('click', () => void open('manual'));
  // Bind explicit settings submission once for the application lifetime.
  if (form) form.addEventListener('submit', save);
  // Bind the session-local pause/resume control.
  node('wellness-pause')?.addEventListener('click', togglePause);
  // Bind the durable stop control.
  node('wellness-stop')?.addEventListener('click', () => void stop());
  // Bind both visible close controls to the same focus-restoring boundary.
  node('wellness-dismiss')?.addEventListener('click', close);
  // Restore focus after native Escape closure as well as explicit close.
  dialog?.addEventListener('close', () => { if (restoreFocus?.isConnected && !restoreFocus.disabled) restoreFocus.focus(); restoreFocus = null; });
  // Listen for hidden-tab return without polling in the background.
  documentRef?.addEventListener('visibilitychange', handleVisibility);

  // Expose only shell lifecycle and deterministic inspection needed by app wiring and tests.
  return Object.freeze({ start, dispose, open, close, localize: render, snapshot: () => ({ enabled: settings.enabled === true, breakEnabled: settings.break_reminder_enabled === true, intervalMinutes: settings.reminder_interval_minutes, paused: localState.paused, lastSlot: localState.lastSlot, pendingVisibleReminder, startedAt, scheduled: timer !== null }) });
}
